### PR TITLE
Add Constant Var support

### PR DESCRIPTION
**Pull request by ChrisJefferson**
This adds to gap the concept of a constant variable. They are created with `MakeConstantVar`, and behave similarly to `MakeReadOnlyVar`, except:

* they can never be changed
* they can only be small integers, true or false
* The interpreter will inline them to optimise byte code.

The inlining is always done, and further to interpreter will inline away `if` statements of the form `if constvar then...` for a constant variable `constvar`.

### CODE REVIEW

File: doc/ref/language.xml [hunk 1]
```diff
@@ -541,7 +541,30 @@ local variables or function arguments).
 <P/>
 First, such variables may be marked <E>read-only</E>. In which case
 attempts to change them will fail. Most of the global variables
-defined in the &GAP; library are so marked.
+defined in the &GAP; library are so marked. <E>read-only</E> variables
+can be made read-write again. GAP also features <E>constant</E> variables,
```
Comment (fingolfin) [inline doc/ref/language.xml]: Perhaps: "can be made read-write again [using the command XYZ]"

File: lib/global.g [hunk 2]
```diff
@@ -72,6 +72,16 @@ IS_READ_ONLY_GLOBAL := IsReadOnlyGVar;
 
 #############################################################################
 ##
+#F  IS_CONSTANT_GLOBAL ( <name> ) determine if a global variable is constant
+##
+##  IS_CONSTANT_GLOBAL ( <name> ) returns true if the global variable
+##  named by the string <name> is constant and false otherwise (the default)
+##
+
+IS_CONSTANT_GLOBAL := IsConstantGVar;
```
Comment (fingolfin) [inline lib/global.g]: Meta-note: It's really crazy that we define kernel function using CamelCase, then define alternate names in SNAKE_CASE, and finally high level names again in CamelCase. (Of course this PR is just imitating what the surrounding code does, so the PR is fine; I just think we may want to resolve this in another PR in the future)

File: src/code.c [hunk 3]
```diff
@@ -977,14 +977,61 @@ void CodeIfEnd (
...
+        PushExpr(cond2);
+        PushStat(body2);
+    }
+    PushExpr(cond1);
     PushStat( body );
 
+    // Some optimisation cases
+    if (nr == 1) {
+        if (TNUM_EXPR(cond1) == T_TRUE_EXPR) {
+            // Leave Statement
```
Comment (fingolfin) [inline src/code.c]: Statement -> statement

File: src/intrprtr.c [hunk 4]
```diff
@@ -2074,6 +2074,31 @@ void            IntrLongFloatExpr (
...
 }
 
+/****************************************************************************
+**
+*F   IntrSmallGAPIntExpr()  . . . . . . .  'interpret' a GAP small integer
+**
+**  'IntrSmallGAPIntExpr' is the action to 'interpret' a existing GAP small
+**  integer. This is used for implementing constants.
+*/
+void IntrSmallGAPIntExpr(Obj val)
```
Comment (fingolfin) [inline src/intrprtr.c]: I really prefer to refer to "immediate integers", as "small integer" is quite ambiguous (a reader may not realize it has a technical meaning). So perhaps `IntrImmediateIntExpr` and `CodeImmediateIntExpr`?

Comment (fingolfin) [reply src/intrprtr.c]: Or `...IntObj`?

File: src/read.c [hunk 5]
```diff
@@ -481,8 +481,23 @@ void ReadCallVarAss (
...
+        if (mode != 'i' && ValGVar(var) && IsConstantGVar(var)) {
+            Obj val = ValAutoGVar(var);
+            if (val == True)
+                IntrTrueExpr();
+            else if (val == False)
+                IntrFalseExpr();
+            else if (IS_INTOBJ(val))
+                IntrSmallGAPIntExpr(val);
+            else
+                SyntaxError("Invalid Constant Variable");
```
Comment (fingolfin) [inline src/read.c]: "Invalid Constant Variable" -> "Invalid constant variable"

File: tst/testinstall/constant.tst [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    fi;
+    if boolfalsevar  then
+        return 4;
+    else
+        return 5;
+    fi;
+    if boolfalsevar  then
+        return 6;
+    fi;
+    return;
```
Comment (fingolfin) [inline tst/testinstall/constant.tst]: Hmm, perhaps also test  `if/elif/fi` and `if/elif/else/fi` ?

**Comment by ChrisJefferson** [comment]
I believe I have fixed issues, and written more tests. This PR includes a commit which fixes printing `elif true`, which was always printed as `else`.

I agree the camel case / allcaps dance is weird (and I would like to remove the methods `MakeConstantGVar` and friends, as to a user I think it's unclear if they should be using `MakeConstantGVar` and `MakeConstantGlobal`).

**Comment by ChrisJefferson** [comment]
Note the optimisations this supports are currently very limited. It only simplifies code of the form `if true then` or `if false then`. We could easily support more advanced simplifications, but I would prefer to add only a small set (a future system based on turning bytecode into GAP objects could do clever, more extensive, rewrites).

**Comment by fingolfin** [comment]
BTW, why do you capitalise "Constant" in e.g. the commit message "Add Constant support" ? Shouldn't it be rather "Add support for constant variables" or so?

File: lib/global.g [hunk 7]
```diff
@@ -92,6 +102,16 @@ MAKE_READ_WRITE_GLOBAL := MakeReadWriteGVar;
 
 #############################################################################
 ##
+#F  MAKE_CONSTANT_GLOBAL ( <name> ) . . .make a global variable constant
```
Comment (fingolfin) [inline lib/global.g]: add at least a space before "make". But actually there should be enough `. . . ` to right-align the text afterwards... Alternatively, just remove the whole ` . . .make a global variable constant` bit, as it is kind of redundant given the following text.

File: lib/global.gd [hunk 8]
```diff
@@ -155,6 +170,26 @@ DeclareGlobalFunction("MakeReadWriteGlobal");
 
 #############################################################################
 ##
+#F  MakeConstantGlobal( <name> )   . . . . make a global variable constant
```
Comment (fingolfin) [inline lib/global.gd]: again the `. . . ` business...  should change the triple spaces to a single one and insert more `. ` -- the same in other places, too, of course

File: lib/global.gd [hunk 9]
```diff
@@ -155,6 +170,26 @@ DeclareGlobalFunction("MakeReadWriteGlobal");
...
+##
+##  <ManSection>
+##  <Func Name="MakeConstantGlobal" Arg='name'/>
+##
+##  <Description>
+##  MakeConstantGlobal ( <A>name</A> ) marks the global variable named
+##  by the string <A>name</A> as constant. A constant variable can never
+##  be changed or made read-write. Constant variables can only take an
+##  integer value, <C>true</C> or <C>false</C>.
```
Comment (fingolfin) [inline lib/global.gd]: should that be "immediate integer value"? Hmm, though not everybody will know what that means..

File: lib/global.gi [hunk 10]
```diff
@@ -199,6 +217,27 @@ InstallGlobalFunction( MakeReadWriteGlobal,
...
 end);
 
+#############################################################################
+##
+#F  MakeConstantGlobal ( <name> )  . . . . make a global variable constant
+##
+##  MakeConstantGlobal ( <name> ) marks the global variable named
+##  by the string <name> as constant
+##
+## A warning is given if <name> is already constant
```
Comment (fingolfin) [inline lib/global.gi]: Add extra space before `A warning`

File: lib/global.gi [hunk 11]
```diff
@@ -199,6 +217,27 @@ InstallGlobalFunction( MakeReadWriteGlobal,
     MAKE_READ_WRITE_GLOBAL( name );
 end);
 
+#############################################################################
+##
+#F  MakeConstantGlobal ( <name> )  . . . . make a global variable constant
+##
+##  MakeConstantGlobal ( <name> ) marks the global variable named
```
Comment (fingolfin) [inline lib/global.gi]: Why the weird wrapping here?

File: doc/ref/language.xml [hunk 12]
```diff
@@ -539,9 +539,34 @@ Certain special facilities are provided for manipulating global
...
 local variables or function arguments).
 <P/>
-First, such variables may be marked <E>read-only</E>. In which case
+First, such variables may be marked <E>read-only</E> using
+<Ref Func="MakeReadOnlyGlobal"/>. In which case
 attempts to change them will fail. Most of the global variables
-defined in the &GAP; library are so marked.
+defined in the &GAP; library are so marked. <E>read-only</E> variables
+can be made read-write again by calling <Ref Func="MakeReadWriteGlobal"/>.
+GAP also features <E>constant</E> variables,
```
Comment (fingolfin) [inline doc/ref/language.xml]: Why the weird wrapping here?

File: doc/ref/language.xml [hunk 13]
```diff
@@ -539,9 +539,34 @@ Certain special facilities are provided for manipulating global
...
+which can never be changed. In some cases, GAP can optimise code which
+uses <E>constant</E> variables, as their value never changes.
```
Comment (fingolfin) [inline doc/ref/language.xml]: Hmm, the user at this point knows very little about constants, and has no idea how to learn more. E.g. they won't know that only immediate integers and some booleans can be made constant. I'd add a ref to `MakeConstantGVar` here, and then make sure they can learn enough about that in its documentation (which currently does not seem to exist...?).

Comment (fingolfin) [reply doc/ref/language.xml]: Getting back to this, it seems this this should be `MakeConstantGlobal` instead (also in the example below)

File: src/code.c [hunk 14]
```diff
@@ -977,14 +977,57 @@ void CodeIfEnd (
...
-    cond = PopExpr();
-    hase = (TNUM_EXPR(cond) == T_TRUE_EXPR);
-    PushExpr( cond );
+    cond1 = PopExpr();
+    hase = (TNUM_EXPR(cond1) == T_TRUE_EXPR);
+    if (nr == 2) {
+        Stat body2 = PopStat();
+        cond2 = PopExpr();
+        PushExpr(cond2);
+        PushStat(body2);
```
Comment (fingolfin) [inline src/code.c]: You don't really need to pop/push `body2` (and the original code also did not need to push/pop `body`...

File: src/code.c [hunk 15]
```diff
@@ -977,14 +977,57 @@ void CodeIfEnd (
...
+        if (TNUM_EXPR(cond1) == T_TRUE_EXPR) {
+            // Leave statement
+            PopExpr();
+            return;
+        }
+        else if (TNUM_EXPR(cond1) == T_FALSE_EXPR) {
+            // Remove entire if statement
+            PopStat();
+            PopExpr();
+            PushStat(NewStat(T_EMPTY, 0));
```
Comment (fingolfin) [inline src/code.c]: Hmm... is that `PushStat` really necessary?

File: src/gvars.c [hunk 16]
```diff
@@ -799,10 +856,9 @@ Obj FuncMakeReadWriteGVar (
 Int IsReadOnlyGVar (
     UInt                gvar )
 {
-    return !INT_INTOBJ(ELM_GVAR_LIST(WriteGVars, gvar));
+    return INT_INTOBJ(ELM_GVAR_LIST(WriteGVars, gvar)) == 0;
```
Comment (fingolfin) [inline src/gvars.c]: Rewrite this as `return ELM_GVAR_LIST(WriteGVars, gvar) == INTOBJ_INT(0)`

File: src/gvars.c [hunk 17]
```diff
@@ -825,6 +881,37 @@ static Obj FuncIsReadOnlyGVar (
...
+**
+*F  FuncIsConstantGVar( <name> ) . . .handler for GAP function
+**
+*/
+
+static Obj FuncIsConstantGVar (
+    Obj                 self,
+    Obj                 name )
+{
+    /* check the argument                                                  */
```
Comment (fingolfin) [inline src/gvars.c]: Personally, I try to avoid those "right aligned with lots of spaces" comments, and just use C++ style `//` comments. While this makes the new code slightly different than the old code, I think we should eventually replace all those weird aligned comments with these

(That's just a vague suggestion, not an actual request to change the PR).

File: doc/ref/language.xml (hunk 12 above)
Comment (alex-konovalov) [reply doc/ref/language.xml]: Also, GAPDoc recommends to use entity `&GAP;` for GAP

File: lib/global.gd (hunk 8 above)
Comment (ChrisJefferson) [reply lib/global.gd]: I copied the triple spaces from the function directly above, and below, this one. This file has either triple spaces, or double spaces, never a single one.

File: lib/global.gi (hunk 11 above)
Comment (ChrisJefferson) [reply lib/global.gi]: once again, same as the wrapping on the function just before

File: lib/global.gi [hunk 18]
```diff
@@ -147,13 +147,31 @@ InstallGlobalFunction( IsReadOnlyGlobal,
...
+##
+##  IsConstantGlobal ( <name> ) returns true if the global variable
+##  named by the string <name> is constant and false otherwise (the default)
+##
+
+InstallGlobalFunction( IsConstantGlobal,
+        function (name)
+    local isro;
+    CheckGlobalName( name );
+    isro := IS_CONSTANT_GLOBAL(name);
```
Comment (fingolfin) [inline lib/global.gi]: Why "isro" ? Probably copy & paste from `IsReadOnlyGlobal`, but this name doesn't make sense here.

File: src/code.c [hunk 19]
```diff
@@ -1616,6 +1654,11 @@ void CodePow ( void )
 }
 
 
+void CodeGAPSmallInt(Obj val)
```
Comment (fingolfin) [inline src/code.c]: I'd really prefer if this was called `CodeImmediateInt` -- saying "small int" is quite ambiguous. This code instead deals with "immediate integer handles" (to use the language from `gmpints.c`)

[ Sorry, I thought I had mentioned this in my previous PR? But perhaps I failed to actually submit that comment? I can't find it now, in any case :-/ ]

File: src/gap.c [hunk 20]
```diff
@@ -2998,12 +2998,20 @@ static Int InitLibrary (
     /* create windows command buffer                                       */
     WindowCmdString = NEW_STRING( 1000 );
 
+    UInt isvar = GVarName( "IsHPCGAP" );
+
 #ifdef HPCGAP
     UInt var = GVarName( "HPCGAP" );
     AssGVar( var, True );
-    MakeReadOnlyGVar( var );
+    MakeConstantGVar( var );
```
Comment (fingolfin) [inline src/gap.c]: Perhaps `IsBound(CONST_VAR)` should be optimized `true` if `CONST_VAR` is bound during coding? (I am not asking for this to be done in this PR, just wondering)

File: src/gvars.c [hunk 21]
```diff
@@ -825,6 +881,34 @@ static Obj FuncIsReadOnlyGVar (
     return IsReadOnlyGVar(GVarName(CSTR_STRING(name))) ? True : False;
 }
 
+/****************************************************************************
+**
+*F  IsConstantGVar( <gvar> ) . . . . . . return if a variable is a constant
+*/
+Int IsConstantGVar(UInt gvar)
+{
+    return INT_INTOBJ(ELM_GVAR_LIST(WriteGVars, gvar)) == -1;
```
Comment (fingolfin) [inline src/gvars.c]: Better (and slightly safer...) code can be generated if you instead write `return ELM_GVAR_LIST(WriteGVars, gvar) == INTOBJ_INT(-1);` -- this would then also match `IsReadOnlyGVar` above.

File: lib/global.gd (hunk 8 above)
Comment (fingolfin) [reply lib/global.gd]: There are 25 places in the whole library using a triple space in that location, i.e. less than 1% -- I think they come from somebody not being careful and/or not caring.

Now, I am first to agree that it sucks to manually update these kinds of comments. But I still think there are only two sane choices: 
1. Abolish them, possibly incrementally (by just not adding them to new code)
2. Keep using them "as intended" -- and then triple spaces to me look quite clearly like a bug, which simply proliferates by copy&paste.

tl;dr: please either delete this line, or format it to match the rest of the library (and not just the rest of this comparatively short file).

File: lib/global.g (hunk 7 above)
Comment (fingolfin) [reply lib/global.g]: Still would like to see that space (or, as in my other comment: Remove that line).

File: src/gvars.c [hunk 22]
```diff
@@ -825,6 +881,34 @@ static Obj FuncIsReadOnlyGVar (
     return IsReadOnlyGVar(GVarName(CSTR_STRING(name))) ? True : False;
 }
 
+/****************************************************************************
+**
+*F  IsConstantGVar( <gvar> ) . . . . . . return if a variable is a constant
```
Comment (fingolfin) [inline src/gvars.c]: Hrmf, somehow GitHub decided that the comments I was collecting for a review were all to be suddenly published as comments instead, taking away from me the possibility to edit them beforehand... I was about to add the following (or possibly, replace those comments completely with it) regarding the comment formatting: "OK, I see now that these files you are touching are particularly bad when it comes to the comment formatting (and grammar...). So it probably makes more sense to go over them separately later on, and to "clean up" those comments, and/or to remove parts of them."

**Comment by stevelinton** [comment]
My only concern with this is that if we want to take it further and have a "proper" optimisation framework then quite a bit of it will be superceded. On reflection though, it would be the implementation, and maybe the printing of functions that had optimised versions that would change, and not the UI. This may be developed further in future -- for instance a Constant Var holding a function would allow inlining. I imagine in future we might also stop printing the optimised versions of functions, but rather hold the original and the optimised bodys and print one while running the other, so again maybe the documentation should make clear that this may change, but we promise it will never hurt to make constant things constant.

**Comment by ChrisJefferson** [comment]
I agree about saying that we don't promise to show optimised functions.

I would prefer a general framework based on @markuspf's transforming of GAP functions to GAP objects -- in particular I would love GAP to have inlining.

Also, I don't think we should expand this much further -- it would be easy to continue through the parser and handle things like `true or X` is true, but then we would end up with effectively "3" execution paths -- immediate execution, turning into bytecode and execution while generating bytecode.

**Comment by fingolfin** [comment]
Supporting simplification of `true or X`, `IsBound(constVar)`, `false and X`, etc. actually seems pretty much straightforward, and I wouldn't consider it a "third execution path" -- OK, in some sens it is, but this is really basic stuff, running in guaranteed constant time. On the plus side, it would make it possible to simply expressions involving multiple constant vars.

Of course it's tricky were to draw the line... Simplifying expressions like `1+1`, or, more realistically, `2^20`, would be a natural next step, and again pretty easy to implement, but of course now things get way more complex, as the result may not fit into an immediate integer, or may even raise an error (think `1/0`). 

And I really think this would already provide some nice (minor) speed up for some code, without needing the full power (and considerable complexity) of a full code transformation framework...

Anyway, that's just some thoughts, but of course they in no way speak against this PR :-).

File: lib/global.gd (hunk 9 above)
Comment (stevelinton) [reply lib/global.gd]: I don't think the concept is defined in the documentation -- not anywhere obvious, anyway. Maybe just say that there is a limit on the size of the integers.

**Comment by alex-konovalov** [comment]
@ChrisJefferson just adding ManSection to the documentation does not automatically add it to the manual - it needs to be hooked somewhere via the `Include` mechanism. This is why cross-reference to `MakeConstantGlobal` is not resolved, as you can see at https://travis-ci.org/gap-system/gap/jobs/279972109. Unfortunately, that test does not fail automatically when cross-references are not resolved (it only fails if the build completely fails to produce manuals), so one should check it manually for any PR that updates manuals.

Perhaps we can detect unresolved cross-references too - but should carefully do that to ignore reports on them during the first run of the build.
